# SKLEARN PROJECT A to Z, let us talk some fundamentals

# Load a built-in dataset from sklearn

In [3]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
iris = load_iris()

## Turn the ds into a dataframe for model training

In [10]:
iris_df = pd.DataFrame(iris["data"],columns=iris["feature_names"])
iris_df["target"]= pd.Series(iris["target"])
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

## Split the data

In [23]:
from sklearn.model_selection import train_test_split
x = iris_df.drop("target",axis=1)
y = iris_df["target"]
np.shape(x),np.shape(y)
X_TRAIN, X_TEST, Y_TRAIN, Y_TEST = train_test_split(x,y,test_size = 0.2)

## Choose and train the model

In [32]:
from sklearn.linear_model import Ridge

# create an object of the Rige class
ridge = Ridge()

# TRAIN ON DATA
ridge.fit(X_TRAIN,Y_TRAIN)

# ACCURACY TRAIN VS TEST
ridge.score(X_TRAIN,Y_TRAIN), ridge.score(X_TEST,Y_TEST)

(0.9634171493354166, 0.9188457972109684)

### TRAINING ACCURACY: 0.9188
### VALIDATION ACCURACY: 0.9634

# SUMMARY

## if you get validation Accuracy slightly higher than training accuracy: it means

## There is potential Regularization during training to prevent overfitting
## Data Augmentation might be applied to make the training task slightly harder resulting in a low score
## There might be differences in metric Calculation i.e. training accuracy is often an avg calcualted over each batch diring an epoch
## Consider for example if the validation set contains simple Samples than the training set
## and a less common issue could be that of underfitting!